Add SENSEI Adaptor for Particles Containers

c-wetterer-nelson opened the pull request:
## Summary
Here, we add two pieces of functionality:
1. The Amr class is currently hard coded to enable SENSEI. This PR adds a new compile flag AMREX_SENSEI_AUTO which allows you to turn off that behavior and then manually instrument an analysis code with SENSEI.

2. The major addition from this PR is a collection of data adaptors and in-situ bridges for adapting Particle Container data via SENSEI. There are two new data adaptors. the `AmrParticleDataAdaptor` and the `AmrParticleMeshDataAdaptor`. The former performs SENSEI data adapting on a ParticleContainer. The latter is a composite adaptor which contains both an AmrParticleDataAdaptor and an AmrDataAdaptor allowing a user to adapt both an Amr object's mesh and the associated Particle Container at the same time. Bridges associated with each of these new data adaptors are also provided.

## Additional background

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [X] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

c-wetterer-nelson committed:
add  AMREX_SENSEI_AUTO compile option

This enables/disables automatic SENSEI instrumentation in Amr

c-wetterer-nelson committed:
add SENSEI particle adaptors/bridges

c-wetterer-nelson commented:
@burlen Let me know what you think!

atmyers commented:
Hi @c-wetterer-nelson,

Thank you for this PR! It looks good to me, what do you think @burlen? 

Would you mind fixing the few minor style issues identified by the "trailing whitespaces" action?

Also, it would be great if you could add a CI action to the amrex-tutorials repo that automatically tested the `SENSEI/Advection_AmrLevel`. This could help us make sure we don't make any breaking changes to SENSEI interoperability on the amrex side.

c-wetterer-nelson committed:
fix trailing whitespaces

burlen commented:
@c-wetterer-nelson I tried to build this today but CMake goes into infinite recursion when building AMReX with `-DAMReX_SENSEI=ON` somewhere in `AMReXGenexHelpers.cmake`. It takes a while but the OOM killer eventually kills the process. I'm using cmake 3.18.4. Should we iterate off line?

print statements in `AMReXGenexHelpers.cmake` show:
```
...
-- get_target_properties_flattened amrex
-- get_target_prop_recursive amrex
-- get_target_prop_recursive Threads::Threads
-- get_target_prop_recursive MPI::MPI_C
-- get_target_prop_recursive MPI::MPI_CXX
-- get_target_prop_recursive sensei
-- get_target_prop_recursive senseiCore
-- get_target_prop_recursive pugixml
-- get_target_prop_recursive thread
-- get_target_prop_recursive sDIY
-- get_target_prop_recursive sVTK
-- get_target_prop_recursive ParaView::Catalyst
-- get_target_prop_recursive VTK::CommonCore
-- get_target_prop_recursive VTK::kwiml
-- get_target_prop_recursive VTK::vtksys
-- get_target_prop_recursive Threads::Threads
-- get_target_prop_recursive VTK::PythonUsed
-- get_target_prop_recursive ParaView::RemotingViews
-- get_target_prop_recursive ParaView::RemotingCore
-- get_target_prop_recursive ParaView::RemotingClientServerStream
-- get_target_prop_recursive VTK::CommonCore
-- get_target_prop_recursive VTK::kwiml
-- get_target_prop_recursive VTK::vtksys
-- get_target_prop_recursive Threads::Threads
-- get_target_prop_recursive ParaView::VTKExtensionsCore
...
```
After what's shown above I see a 100s/1000s of  `VTK::X` scrolling by until it's killed. I added the full output in an attachment

[log.txt](https://github.com/AMReX-Codes/amrex/files/6496217/log.txt)

c-wetterer-nelson commented:
Hi @burlen, that is a very weird behavior. I don't see anything out of the ordinary upstream of that forever block of get_target_prop_recursive statements. My hunch is that it has something to do with the ParaView/VTK libs that SENSEI is linked against ahead of linking SENSEI into amrex. What do your CMake options look like going into amrex?

Can you try turning off AMReX_SENSEI_AUTO, leaving AMReX_SENSEI on? I added that option in this PR, and it defaults to ON in order to preserve the previous behavior of SENSEI here.

c-wetterer-nelson commented:
@burlen, I'm also curious, did you attempt here to compile amrex into a fresh build directory, or just build on top of an existing build tree?

burlen commented:
These are all fresh builds. I've used cmake long enough to know better than attempt to use it on something existing. 

My command line:
```
cmake -DAMReX_SENSEI=ON -DSENSEI_DIR=/work/SENSEI/sensei-cwn-install-catalyst/lib/cmake/  ..
```

I was able to get past the infinite recursion mentioned above by patching the cmake code.
After cmake ran, I had to explicitly edit the CMakeCache.txt and change `-std=c++11` to `-std=c++14`.
I also had to patch `AMReX_Amr.H`  to fix a typo in an ifdef

with that the build finishes at least! There may be more issue when we try to use the library though.

burlen committed:
fix a typo in ifdef

burlen committed:
avoid inifinte recursion when linking ParaView Catalyst

burlen commented:
@c-wetterer-nelson could you give me permission to push to your fork?

c-wetterer-nelson commented:
ah great!

burlen committed:
cmake export variables used in build config

burlen committed:
Use the correct version of SENSEI when importing the CMake config

* Adds the path to the SENSEI install used during the build into the
  CMake config so that it's not necessary to do so when linking to
  libamrex.

burlen committed:
fix the Makefile build with SENSEI

* Adds the new particle adaptor classes
* Updates the macros and options.

burlen commented:
@atmyers @c-wetterer-nelson This is great! I am putting it through its paces in combination with the tutorial PR. I have a few build issues and Corey and I are iterating on these now. I'll post back when those are resolved.

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.H:
```diff
@@ -0,0 +1,100 @@
+int AmrParticleInSituBridge<NStructReal, NStructInt, NArrayReal, NArrayInt>::update(
+  amrex::Amr *amr,
+  amrex::ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt> *tracers,
+  std::map<std::string, std::vector<int>> particles_rStructs,
+  std::map<std::string, int> particles_iStructs,
+  std::map<std::string, std::vector<int>> particles_rArrays,
+  std::map<std::string, int> particles_iArrays)
```
these should probably be passed by reference to const

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleDataAdaptor.H:
```diff
@@ -0,0 +1,103 @@
+  vtkPolyData* BuildParticles();
+
+  const std::string m_particlesName = "particles";
+  amrex::ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt>* m_tracers;
```
should this be named "m_particles"?

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.H:
```diff
@@ -0,0 +1,100 @@
+  int update(
+    amrex::Amr *amr,
+    amrex::ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt> *tracers,
+    std::map<std::string, std::vector<int>> particles_rStructs = {},
+    std::map<std::string, int> particles_iStructs = {},
+    std::map<std::string, std::vector<int>> particles_rArrays = {},
+    std::map<std::string, int> particles_iArrays = {});
```
these should probably be passed as reference to const

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.H:
```diff
@@ -0,0 +1,100 @@
+{
+/// Bridge for simulations that use amrex::ParticleContainer
+/// inherits intialize and finalize from amrex::InSituBridge
+template<int NStructReal, int NStructInt, int NArrayReal, int NArrayInt>
+class AmrParticleInSituBridge : public InSituBridge
```
since there are no member variables that depend on the template parameters, the template should be on the update method not the class. this will make the class easier to use.

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.cpp:
```diff
@@ -0,0 +1,8 @@
+#include <AMReX_AmrParticleInSituBridge.H>
+
+#ifdef AMREX_PARTICLES
+namespace amrex {
+// explicit template specialization for use with AmrParticleContainer class
+template<> class AmrParticleInSituBridge<3,0,0,0>;
+}
```
what's this used for?

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.H:
```diff
@@ -0,0 +1,100 @@
+/// Bridge for simulations that use amrex::ParticleContainer
+/// inherits intialize and finalize from amrex::InSituBridge
+template<int NStructReal, int NStructInt, int NArrayReal, int NArrayInt>
+class AmrParticleInSituBridge : public InSituBridge
```
the naming convention in use is "[AMReX class]InSituBridge" and the AMReX class here is Particle (or ParticleContainer), we could drop the Amr. This will be handy since we need to handle cases where there's both a mesh and particles in which the mesh could come from either an amrex::Amr instance or and amrex::AmrMesh instances, and those classes would be included in the bridge & adaptor names.

c-wetterer-nelson replied:
I'm with you here. I just added a change in 
[!285e83a](https://github.com/AMReX-Codes/amrex/pull/2016/commits/285e83aa78a970fe19e4072d170d41c071244529) to refine the naming convention. now we have:

ParticleDataAdaptor: exclusively processes a Particle Container
AmrParticleDataAdaptor: the composite adaptor, processes an Amr object and a Particle Container

ParticleInSituBridge: exclusively processes a Particle Container
AmrParticleInSituBridge: the composite adaptor, processes an Amr object and a Particle Container

This now leaves room for another composite data adaptor/bridge which will process an AmrMesh + Particle Container. This will be called
AmrMeshParticleDataAdaptor and AmrMeshParticleInSituBridge

burlen committed:
SENSEI build options and macros

* BL_USE_SENSEI_INSITU upgraded to AMREX_USE_SENSEI_INSITU
* AMREX_USE_SENSEI_AUTO becomes AMREX_NO_SENSEI_AMR_INST so that if the
  macro is not defined the old behavior of enabling the internal
  instrumentation in amrex::Amr is in effect.

burlen commented on Src/Extern/SENSEI/AMReX_AmrParticleInSituBridge.H:
```diff
@@ -0,0 +1,100 @@
+  // invoke the in situ analysis with data from an AmrMesh
+  // instance.
+  int update(
+    amrex::Amr *amr,
```
This bridge does not have the capability of visualizing the data in the amrex::Amr object. We are are getting time & time step from the passed instance. Instead of an amrex::Amr instance we can pass time and time step directly, so the bridge need not depend on amrex::Amr. Note, in some of the particle tutorials for example amrex::Amr is not used.

c-wetterer-nelson committed:
make array name maps const refs

c-wetterer-nelson committed:
swap 'tracers' for 'particles'

c-wetterer-nelson committed:
remove amr from particle exclusive bridge

c-wetterer-nelson committed:
refactor templating for particle bridges

c-wetterer-nelson committed:
refine adaptor naming convention

burlen committed:
clean up warnings in SENSEI codes

burlen committed:
fix a syntax error in the SENSEI ParticleInSituBridge

burlen committed:
fix a syntax error in the SENSEI AmrParticleInSituBridge

burlen committed:
SENSEI ParticleDataAdaptor SINGLE_PRECISION_PARTICLES and neighbor particles

* use loop bounds from numRealParticles rather than using the iterators
  as the iterators iterate neighbor particles which shouldn't be
  trasnfered.
* deal with compile time particle precision selection

burlen committed:
SENSEI particle data adaptor metadata fixes

* handle compile time precision selection
* report per block bounds.
* issue a warning if block array ranges are requested, since that's
  not yet implemented.

burlen commented:
@atmyers I think this is finally ready for a review. I've rebased on your development branch as of this am, should be go to merge, if you're OK with the changes.

atmyers commented on Src/Extern/SENSEI/AMReX_AmrDataAdaptor.cpp:
```diff
@@ -605,6 +605,8 @@ int AmrDataAdaptor::GetArrayName(const std::string &meshName,
 int AmrDataAdaptor::GetMesh(const std::string &meshName,
     bool structureOnly, vtkDataObject *&mesh)
 {
+    (void) structureOnly;
```
```suggestion
    amrex::ignore_unused(structureOnly);
```

This is maybe slightly more clear that the intent is to silence unused variable warnings.

atmyers commented on Src/Extern/SENSEI/AMReX_AmrDataAdaptor.cpp:
```diff
@@ -817,6 +819,9 @@ int AmrDataAdaptor::AddGhostCellsArray(vtkDataObject* mesh,
 int AmrDataAdaptor::AddGhostNodesArray(vtkDataObject *mesh,
     const std::string &meshName)
 {
+    (void) mesh;
```
```suggestion
    amrex::ignore_unused(mesh);
```

atmyers commented on Src/Extern/SENSEI/AMReX_AmrDataAdaptor.cpp:
```diff
@@ -817,6 +819,9 @@ int AmrDataAdaptor::AddGhostCellsArray(vtkDataObject* mesh,
     const std::string &meshName)
 {
+    (void) mesh;
+    (void) meshName;
```
```suggestion
    amrex::ignore_unused(meshName);
```

atmyers commented on Src/Extern/SENSEI/AMReX_AmrInSituBridge.H:
```diff
@@ -9,8 +9,7 @@ namespace amrex
 
-/// Bridge for simulations that use amrex::Amr
-/// inherits intialize and finalize from amrex::InSituBridge
+/// Conatins the bridge code for simulations that use amrex::Amr
```
```suggestion
/// Contains the bridge code for simulations that use amrex::Amr
```

atmyers commented on Src/Extern/SENSEI/AMReX_AmrMeshDataAdaptor.cpp:
```diff
@@ -477,6 +477,8 @@ int AmrMeshDataAdaptor::GetMeshHasGhostCells(const std::string &meshName, int &n
 int AmrMeshDataAdaptor::GetMesh(const std::string &meshName,
     bool structureOnly, vtkDataObject *&mesh)
 {
+    (void) structureOnly;
```
```suggestion
    amrex::ignore_unused(structureOnly);
```

atmyers commented on Src/Extern/SENSEI/AMReX_AmrMeshDataAdaptor.cpp:
```diff
@@ -598,6 +600,9 @@ int AmrMeshDataAdaptor::GetMesh(const std::string &meshName,
 int AmrMeshDataAdaptor::AddGhostNodesArray(vtkDataObject *mesh,
     const std::string &meshName)
 {
+    (void) mesh;
```
```suggestion
    amrex::ignore_unused(mesh);
```

atmyers commented on Src/Extern/SENSEI/AMReX_AmrMeshDataAdaptor.cpp:
```diff
@@ -598,6 +600,9 @@ int AmrMeshDataAdaptor::GetMesh(const std::string &meshName,
     const std::string &meshName)
 {
+    (void) mesh;
+    (void) meshName;
```
```suggestion
    amrex::ignore_unused(meshName);
```

atmyers commented on Src/Extern/SENSEI/AMReX_InSituUtils.H:
```diff
@@ -98,6 +98,8 @@ template <typename n_t>
 int AllocateBoxArray(const amrex::Box &pdom, const amrex::BoxArray &boxes,
     const amrex::DistributionMapping &dmap, int ng, std::vector<n_t*> &darrays)
 {
+    (void) pdom;
```
```suggestion
    amrex::ignore_unused(pdom);
```

atmyers commented on Src/Extern/SENSEI/AMReX_InSituUtils.H:
```diff
@@ -154,6 +156,8 @@ int MaskGhostCells(const amrex::Box &pdom, const amrex::BoxArray &boxes,
     const amrex::DistributionMapping &dmap, unsigned int ng,
     std::vector<n_t*> &darrays)
 {
+    (void) pdom;
```
```suggestion
    amrex::ignore_unused(pdom);
```

atmyers commented on Src/Extern/SENSEI/AMReX_InSituUtils.H:
```diff
@@ -234,6 +238,8 @@ int MaskCoveredCells(const amrex::Box &pdom, const amrex::BoxArray &cBoxes,
     const amrex::DistributionMapping &cMap, amrex::BoxArray fBoxes,
     const IntVect &fRefRatio, int ng, std::vector<n_t*> &cMasks)
 {
+    (void) pdom;
```
```suggestion
    amrex::ignore_unused(pdom);
```

atmyers commented:
Thank you for this PR - it looks great to me. I made a only a few minor cosmetic suggestions.

burlen committed:
clarify intent to silence unused variable warnings.

Co-authored-by: Andrew Myers <atmyers2@gmail.com>

burlen committed:
clarify intent to silence warnings

Co-authored-by: Andrew Myers <atmyers2@gmail.com>

burlen committed:
clarify intent to silence unused variable warnings.

Co-authored-by: Andrew Myers <atmyers2@gmail.com>
